### PR TITLE
Refactor distribution and extend tests for sub-distributions

### PR DESCRIPTION
**Pull request by rasolca**
This PR consists of
- util_distribution.h
   - [x] refactoring
   - [x] extend tests with offsets
     - [x] fix a couple of bugs when using offsets
- distribution.h
  - [x] refactoring
  - [x] extract specific methods to distribution_extensions.h (internal methods)
  - [x] create aliases with the old names (which can be deprecated)
  - [x] extend tests
    - [x] fix bugs (Note: some bugs already fixed but not tested.)
  - [x] implement new methods needed for the distributed tridiagonal solver in distribution_extensions.h

### CODE REVIEW

**Review by msimberg — COMMENTED** [review]
It's a bit hard to follow the changes because of the renamings, but from what I can tell it looks all valid. Thanks for spotting these issues!

I only have a few higher level comments at this point.

File: include/dlaf/matrix/distribution.h [hunk 1]
```diff
@@ -148,416 +182,386 @@ class Distribution {
...
+    return global_element_from_global_tile_and_tile_element<rc>(global_tile, tile_element);
   }
 
-  /// Returns the rank index of the process that stores the tile with global index @p global_tile.
+  template <Coord rc>
+  SizeType local_element_from_global_element(const SizeType global_element) const noexcept {
+    const auto global_tile = global_tile_from_global_element<rc>(global_element);
+    const auto local_tile = local_tile_from_global_tile<rc>(global_tile);
+    const auto tile_element = tile_element_from_global_element<rc>(global_element);
+    return local_element_from_local_tile_and_tile_element<rc>(local_tile, tile_element);
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix/distribution.h:353]: Sorry, the implementation changes were not meant to be review yet.
Just the interface.

File: include/dlaf/matrix/distribution.h [hunk 2]
```diff
@@ -17,8 +17,42 @@
...
 #include <dlaf/util_math.h>
 
-namespace dlaf {
-namespace matrix {
+#define DLAF_DISTRIBUTION_ENABLE_DEPRECATED 0
+#if (DLAF_DISTRIBUTION_ENABLE_DEPRECATED)
+#define DLAF_DISTRIBUTION_DEPRECATED(x) [[deprecated(x)]]
+#else
+#define DLAF_DISTRIBUTION_DEPRECATED(x)
+#endif
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix/distribution.h:25]: This is mainly for me. In this PR I will not update all the algorithms (mainly to not disrupt other people work), but I would do them step by step, and the warnings are helpful to check that everything was done.

File: include/dlaf/matrix/distribution.h [hunk 3]
```diff
@@ -148,416 +182,386 @@ class Distribution {
...
 
-  const TileElementSize& blockSize() const noexcept {
+  const TileElementSize& block_size() const noexcept {
     return block_size_;
   }
 
-  const TileElementSize& baseTileSize() const noexcept {
+  const TileElementSize& tile_size() const noexcept {
     return tile_size_;
   }
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix/distribution.h:205]: My though was:
`tile_size` -> returns tile_size_ member
`size_of_tile(ij)` -> returns the size of the tile ij.

**Comment by rasolca** [comment]
> It's a bit hard to follow the changes because of the renamings, but from what I can tell it looks all valid. Thanks for spotting these issues!
> 
> I only have a few higher level comments at this point.

Sorry, the first review was only meant to check if the new interface made sense.

File: include/dlaf/matrix/distribution.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/matrix/distribution.h:25]: Fair enough, appreciated!

File: include/dlaf/matrix/distribution.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/matrix/distribution.h:205]: To be clear, I found the _name_ of `size_of_tile` weird, not its existence. Sorry if that was unclear. I see the use case for both, I just think saying the same thing in two different ways isn't a good way to distinguish the two use cases. For what it's worth, they could be named exactly the opposite and it'd make as much sense. I don't think the `tile_size()` -> `tile_size_` member connection is strong enough motivation to call only that one `tile_size`.

- `base_tile_size`: returns `tile_size_` member
- `tile_size(ij)`: returns size of tile ij
?

Alternatives to the first:
- `max_tile_size`
- `full_tile_size` (to distinguish between an "incomplete" and "full" tile?)
- ???

[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix/distribution.h:205]: I think my comment was not clear. I know that we agree on having the two methods. I was just explained how I came out with the new names too see if the second name could be improved.

The second method returns the size of the tile with index ij -> in a shorter way: returns the size of tile ij -> size_of_tile(GlobalTileIndex ij)

However I would like to keep `tile_size()` for the first case to be consistent:
- the member is called `tile_size_`
- in the constructors the parameter is named tile_size (and the doc always mention tile_size)
- naming should similar to block_size (block_size has the same problem as the first and the last might be incomplete)

[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix/distribution.h:205]: After a discussion with @msimberg and @albestro we decided for
`tile_size()` and `tile_size_of(ij)`.

File: include/dlaf/matrix/distribution_extensions.h [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+template <Coord rc>
+SizeType distance_to_adjacent_tile(const Distribution& dist, SizeType global_element) noexcept {
+  const SizeType global_tile = dist.global_tile_from_global_element<rc>(global_element);
+  const SizeType tile_element = dist.tile_element_from_global_element<rc>(global_element);
+  return dist.size_of_tile<rc>(global_tile) - tile_element;
+}
+
+/// Returns a local linear column-major index of the tile that contains @p ij_local
+inline SizeType local_tile_linear_index(const Distribution& dist, LocalTileIndex ij_local) noexcept {
+  return ij_local.row() + ij_local.col() * dist.local_nr_tiles().rows();
```
Comment (albestro) [inline include/dlaf/matrix/distribution_extensions.h:100]: Using distribution I realised that an assert here would have helped.

```cpp
DLAF_ASSERT(ij_local.isIn(dist.local_nr_tiles()), ij_local, dist.local_nr_tiles());
```

Probably worth having this kind of assert in here more extensively?

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

File: src/matrix/distribution.cpp [hunk 5]
```diff
@@ -98,85 +100,61 @@ Distribution::Distribution(Distribution rhs, const SubDistributionSpec& spec)
...
+
+    // Fix last tile size
+    if (rank_index_.col() == rank_global_tile<Coord::Col>(nr_tiles_.cols() - 1))
+      // remove the elements missing in the last tile
+      col -= nr_tiles_.cols() * tile_size_.cols() -
+             (size_.cols() + global_tile_element_offset<Coord::Col>());
   }
 
   // Set local_size_.
   local_size_ = LocalElementSize(row, col);
```
Comment (msimberg) [inline src/matrix/distribution.cpp:154]: Would it make sense to factor out a helper which is templated on the `Coord` just to avoid repeating the same code for `Row` and `Col`?

File: test/unit/matrix/test_distribution.cpp [hunk 6]
```diff
@@ -248,392 +393,123 @@ struct ParametersIndices {
...
-
-TEST(DistributionTest, LocalElementDistanceFromGlobalTile) {
-  for (const auto& test : tests_local_distance_from_global_tile) {
-    Distribution obj(test.size, test.block_size, test.grid_size, test.rank, test.src_rank, test.offset);
-
-    EXPECT_EQ(test.local_element_distance,
-              obj.localElementDistanceFromGlobalTile(test.global_tile_begin, test.global_tile_end));
-  }
-}
-
```
Comment (msimberg) [inline test/unit/matrix/test_distribution.cpp:636]: Since it's hard to tell from the diff: Did these test cases and tests get merged into other tests, are they missing, or are they no longer needed (covered by other tests)?

**Review by msimberg — APPROVED** [review]
This looks good to me! Just a couple of general comments, not blocking.

File: test/unit/matrix/test_distribution.cpp (hunk 6 above)
Comment (rasolca) [reply test/unit/matrix/test_distribution.cpp:636]: 341-488: Moved to `test_distribution_extensions.cpp`
489-636: The method tested doesn't exists anymore (the 2D method was never used).
         I can refactor it for the 1D method, but I don't want to spend time for
         something that will disappear after the tridiagonal_eigensolver Q4 clean-up.

**Comment by rasolca** [comment]
cscs-ci run

File: test/unit/matrix/test_distribution.cpp (hunk 6 above)
Comment (msimberg) [reply test/unit/matrix/test_distribution.cpp:636]: All good, thanks for the explanation! I just wanted to make sure they hadn't been accidentally lost.

**Review by msimberg — APPROVED** [review]

**Comment by rasolca** [comment]
cscs-ci run

File: include/dlaf/matrix/distribution.h [hunk 7]
```diff
@@ -148,504 +226,786 @@ class Distribution {
...
+  LocalTileIndex localTileIndex(const GlobalTileIndex& global_tile) const {
+    return local_tile_index(global_tile);
+  }
+
+  DLAF_DISTRIBUTION_DEPRECATED("method has been renamed in snake case")
+  TileElementIndex tileElementIndex(const GlobalElementIndex& global_element) const noexcept {
+    return tile_element_index(global_element);
+  }
+
+  TileElementSize tileSize(const GlobalTileIndex& index) const noexcept {
```
Comment (albestro) [inline include/dlaf/matrix/distribution.h:808]: Why isn't this going to be deprecated? Shouldn't we use `tile_size_of` in the future?

File: include/dlaf/matrix/distribution.h [hunk 8]
```diff
@@ -148,504 +226,786 @@ class Distribution {
...
+    return matrix::internal::distribution::distance_to_adjacent_tile<rc>(*this, global_element);
+  }
+
+  // tridsolver
+  DLAF_DISTRIBUTION_DEPRECATED("Use method from include/dlaf/matrix/distribution_extensions.h")
+  SizeType localTileLinearIndex(LocalTileIndex ij_local) const noexcept {
+    return matrix::internal::distribution::local_tile_linear_index(*this, ij_local);
+  }
+
+  // tridsolver, permutation, test
```
Comment (albestro) [inline include/dlaf/matrix/distribution.h:950]: Just wondering if it is worth opening an issue for these comments here (just one for all, referring the code) to keep track of the cleanup work to be done.

**Review by albestro — APPROVED** [review]
Had a quick look, LGTM.

File: include/dlaf/matrix/distribution.h (hunk 8 above)
Comment (rasolca) [reply include/dlaf/matrix/distribution.h:950]: Added to #1006

**Comment by rasolca** [comment]
cscs-ci run
